crates/bevy_render/macros/src/as_bind_group.rs

```diff
@@ -385,5 +393,451 @@ pub fn derive_as_bind_group(input: TokenStream) -> TokenStream {
                 })
             }
         }
+    }))
+}
+
+/// Represents the arguments for the `uniform` binding attribute.
+///
+/// If parsed, represents an attribute
+/// like `#[uniform(LitInt, Ident)]`
+struct UniformBindingMeta {
+    lit_int: LitInt,
+    _comma: Token![,],
+    ident: Ident,
+}
+
+/// Represents the arguments for any general binding attribute.
+///
+/// If parsed, represents an attribute
+/// like `#[foo(LitInt, ...)]` where the rest is optional `NestedMeta`.
+enum BindingMeta {
+    IndexOnly(LitInt),
+    IndexWithOptions(BindingIndexOptions),
+}
+
+/// Represents the arguments for an attribute with a list of arguments.
+///
+/// This represents, for example, `#[texture(0, dimension = "2d_array")]`.
+struct BindingIndexOptions {
+    lit_int: LitInt,
+    _comma: Token![,],
+    meta_list: Punctuated<NestedMeta, Token![,]>,
+}
+
+impl Parse for BindingMeta {
+    fn parse(input: ParseStream) -> Result<Self> {
+        if input.peek2(Token![,]) {
+            input.parse().map(Self::IndexWithOptions)
+        } else {
+            input.parse().map(Self::IndexOnly)
+        }
+    }
+}
+
+impl Parse for BindingIndexOptions {
+    fn parse(input: ParseStream) -> Result<Self> {
+        Ok(Self {
+            lit_int: input.parse()?,
+            _comma: input.parse()?,
+            meta_list: input.parse_terminated(NestedMeta::parse)?,
+        })
+    }
+}
+
+impl Parse for UniformBindingMeta {
+    fn parse(input: ParseStream) -> Result<Self> {
+        Ok(Self {
+            lit_int: input.parse()?,
+            _comma: input.parse()?,
+            ident: input.parse()?,
+        })
+    }
+}
+
+fn get_uniform_binding_attr(attr: &syn::Attribute) -> Result<(u32, Ident)> {
+    let uniform_binding_meta = attr.parse_args_with(UniformBindingMeta::parse)?;
+
+    let binding_index = uniform_binding_meta.lit_int.base10_parse()?;
+    let ident = uniform_binding_meta.ident;
+
+    Ok((binding_index, ident))
+}
+
+fn get_binding_nested_attr(attr: &syn::Attribute) -> Result<(u32, Vec<NestedMeta>)> {
+    let binding_meta = attr.parse_args_with(BindingMeta::parse)?;
+
+    match binding_meta {
+        BindingMeta::IndexOnly(lit_int) => Ok((lit_int.base10_parse()?, Vec::new())),
+        BindingMeta::IndexWithOptions(BindingIndexOptions {
+            lit_int,
+            _comma: _,
+            meta_list,
+        }) => Ok((lit_int.base10_parse()?, meta_list.into_iter().collect())),
+    }
+}
+
+#[derive(Default)]
+enum ShaderStageVisibility {
+    #[default]
+    All,
```

Comment:
The default previously was to have all visibility, but no material uses compute so this sounds reasonable.

Reply:
It would mean that "by default" AsBindGroup won't work with compute shaders, which is a scenario we probably want to support (and increasingly so as the renderer progresses).

Its hard to have this conversation without knowing the performance tradeoffs. When I ran benchmarks of All vs Fragment (for the pbr shader), I couldn't identify any differences. But that doesn't necessarily mean that this is true for all platforms / scenarios.

Reply:
Best I could find was kvark responding with "could be, yes" to this question:
> When choosing bind group visibility, is there any performance implications of choosing something more specific like ShaderStages::VERTEX over just all?

Reply:
Short term, I'm cool with limiting this to vertex|fragment given that compute is still pretty niche. But we should revisit this decision when it becomes more relevant.

Reply:
I was also thinking that aside from performance, enabling a resource for compute usage could change something about some validation/usage constraints as well. I don't know.